### ISSUE TITLE
`prefer_interpolation_to_compose_strings` range for the string chain

### ISSUE THREAD

**Issue by asashour** [dart-lang/linter#3814]
**Describe the issue**
When the strings are more than two, the range should cover all, not only the first two, for [prefer_interpolation_to_compose_strings](https://dart-lang.github.io/linter/lints/prefer_interpolation_to_compose_strings.html)

The quick fix handles it, but the range reported by the lint should be changed.

**To Reproduce**
```dart
void f(String x) {
  var a = x + '' + x;
  var b = '' + '' + '';
}
```

**Expected behavior**
The range for both cases cover the three `Strings`

**Actual Behavior**
<img src=https://user-images.githubusercontent.com/2410127/199945885-047354ed-9eea-4291-85a0-5eb6be97d745.png width=250>
Only the second case covers the three parts, but not the first one.

**Comment by bwilkerson** [comment]
We generally highlight the smallest range that still gives the user enough context to understand the problem. It isn't clear to me that highlighting all of the operands is better for that purpose than just highlighting the first two.

Also, is it really `prefer_interpolation_to_compose_strings` that's highlighting the second line? I would have thought that was a different lint.

**Comment by srawlins** [comment]
Both `x + '' + x` and `'' + '' + ''` are reported by this rule. The reason for why the longer range, with two binary expressions, is reported in the second case is likely accidental, just how the rule works:

* In `x + '' + x`, it checks the first binary expression (`(x + '') + x`), and sees that neither operand is a string literal, SKIP. Then it examines the second binary expression (`x + ''`), and reports lint.
* In `'' + '' + ''`, it checks the first binary expression (`('' + '') + ''`), and sees that the second operand is a string, REPORT.

if we walk binary expressions DFS-style, it should remove the inconsistency.

**Comment by lrhn** [comment]
> Also, `'' + ''` is considered OK by the comments and tests

Why? It may be "safe" for allocation, because `String.operator+` does nothing if one of the operands is the empty string, but it's still unnecessary code.
Dart does not coerce on string addition, like Java and JavaScript do, so `'' + expr` is just writing `expr` in an unnecessarily complicated way.

**Comment by srawlins** [comment]
> Why?

I have no idea. In the [very first commit](https://github.com/dart-lang/linter/commit/e6dc33b91e1adcd7584786577911c19ceb49c3d8#diff-6b8fcc72bdaad43f218e4841fdd255fb27507577775394acb94025a89e59cc55) that adds this rule, there is a test case line:

```dart
String name = 'Juan'+'Pablo'; // OK // It is not the business of this rule
```

🤷

**Comment by lrhn** [comment]
Ah, so it's adding precisly two strings, not adding *empty* strings, that is ignored. That makes sense.
There is nothing inefficient about `'Juan'+'Pablo'`. It should be formatted though :)
Carry one then :wave:

**Comment by srawlins** [comment]
Fixed by https://github.com/dart-lang/sdk/commit/076dd1a2c07ef8a6d4312ce396e48a98cb77f4d0
